shared/js/content-scripts/block-cookie.js

```diff
@@ -28,72 +28,109 @@
-                        comps[i] = 'max-age=' + nextWeek
-                    }
-
-                    break
+        // Listen for a message from the content script which will configure the policy for this context
+        const loadPolicy = new Promise((resolve) => {
+            const messageListener = (event) => {
+                if (event && event.isTrusted && event.data && event.data.source === secret) {
+                    resolve(event.data)
+                    window.removeEventListener('message', messageListener)
                 }
             }
-
-            if (!expiryFound) {
-                comps.push('max-age=' + nextWeek)
+            window.addEventListener('message', messageListener)
+        })
+        console.log('inject cookie test', cookieSetter, secret)
+        document.__defineSetter__('cookie', (value) => {
+            // call the native document.cookie implementation. This will set the cookie immediately
+            // if the value is valid. We will override this set later if the policy dictates that
+            // the expiry should be changed.
+            cookieSetter.apply(document, [value])
+            try {
+                // determine the origins of the scripts in the stack
+                const scriptOrigins = (new Error().stack.split('\n')).reduce((origins, line) => {
+                    const res = line.match(lineTest)
+                    if (res && res[2]) {
+                        origins.add(new URL(res[2]).hostname)
+                    }
+                    return origins
+                }, new Set())
+                // wait for config before doing same-site tests
+                loadPolicy.then(({ tabRegisteredDomain, policy }) => {
+                    const sameSiteScript = [...scriptOrigins].every((host) => host === tabRegisteredDomain || host.endsWith(`.${tabRegisteredDomain}`))
+                    if (sameSiteScript) {
+                        // cookies set by scripts loaded on the same site as the site are not modified
+                        console.log('ignored', value, scriptOrigins)
+                        return
+                    }
+                    // extract cookie expiry from cookie string
+                    const cookieParts = value.split(';')
+                    const maxAgeIdx = cookieParts.findIndex(part => part.trim().toLowerCase().startsWith('max-age'))
+                    const expiresIdx = cookieParts.findIndex(part => part.trim().toLowerCase().startsWith('expires'))
+                    if (maxAgeIdx === -1 && expiresIdx === -1) {
+                        // session cookie
+                        return
+                    }
+                    const expiry = maxAgeIdx >= 0
+                        ? parseInt(cookieParts[maxAgeIdx].split('=')[1])
+                        : (new Date(cookieParts[expiresIdx].split('=')[1]) - new Date()) / 1000
+                    // apply cookie policy
+                    if (expiry > policy.threshold) {
```

Comment:
to be clear, we are modifying the expiry even if page is safelisted, correct? Let's note this in the tech design (I know you proposed it in the comments).

Reply:
Yes, this policy applies _always_, though with this implementation it would be easy to skip this on pages that the user has allowed.

Reply:
Resolving since we discussed it in Tech Design and changed direction to skip safelisted pages for now.